enhancements/update/upgrades-blocking-on-ack.md

```diff
@@ -0,0 +1,210 @@
+---
+title: upgrade-acknowledgment-gate
+authors:
+  - "@jottofar"
+reviewers:
+  - "@LalatenduMohanty"
+  - "@bparees"
+  - "@sdodson"
+  - "@csrwng"
+  - "@relyt0925"
+  - "@jeremyeder"
+approvers:
+  - TBD 
+creation-date: 2021-07-08
+last-updated: 2021-07-14
+status: implementable
+---
+
+# Upgrade Acknowledgment Gate
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+CVO currently blocks minor level upgrades when overrides are set, a resource deletion is pending, or an operator has set its upgradeable condition to false.
+This enhancement adds a generic upgrade blocking mechanism for blocking minor level upgrades.
+This generic upgrade blocker can consist of one or more items each of which will block a minor level upgrade.
+Each generic upgrade blocker item will be associated with some well-defined task to be completed by an administrator.
+Once the task has been completed the administrator will acknowledge completion and its associated item will no longer block minor level upgrades.
+
+## Motivation
+
+OCP 4.9 uses kubernetes v1.22 which is removing a large number of [APIs][k8s-1.22-removed-apis].
+We suspect that many customer clusters are still using these APIs through helm charts, operators, generic workloads, administrative tools, etc.
+We believe there is no effective way for a Cluster Operator to detect the use of these APIs by external applications.
+Relying solely on automated API detection to block upgrades could result in blocking for APIs that are not actively used any longer or not blocking because an application was not active during the check.
+Having a human in the loop to drive the analysis and make the final decision is the surest way to prevent customers from upgrading from 4.8.z to 4.9.z and encountering missing API related breakage.
+
+### Goals
+
+* Acknowledgment requires the explicit update of a cluster resource by an administrator.
+* For the initial release in 4.8.z, administrator acknowledgment of API removal will be required before a minor level upgrade to 4.9 is allowed.
+* Upgrade blocking mechanism must be generic such that it can be repurposed in the future.
+* It must be clear what an administrator is acknowledging.
+* Administrators can bypass the upgrade blocking mechanism by using the existing CVO "force" implementation.
+* The UI will clearly indicate when this upgrade blocking mechanism is blocking upgrades, why, and what needs to be done to clear the block.
+* This upgrade blocking mechanism must be compatible with SD, ACM, IKS/ROKS, Hive, and Hypershift.
```

Comment:
When I think of SD compatibility I have a key concern: how do we keep customers from directly manipulating this configuration?  If this is part of a new CRD then RBAC can be managed better and webhooks created to block operations from cluster-admins if needed on OSD, ROSA, and ARO.  If this configuration is included in something already existing like ClusterVersion it becomes difficult to cordon permissions from customer.  If it is something generic like a configmap it may be OK if it's a specific name and we never permit customers to create that resource.  This isn't just in protecting the blocking of edges but also from keeping customers from setting the block manually.  Not clear how this will be used in the future so I have to assume behavior can change, therefore I'd like clear ways to manage permissions for users. 

So a part of this should be SD can easily prevent customers from directly manipulating this configuration.

Reply:
the two configmaps in this EP will have well defined names and live in a well defined namespace.  Does that satisfy your concerns?

(one of them defines the gates and should never be edited by an admin at all..if they try, the CVO is going to stomp the change back anyway, the other is where the admin puts their ack of the gate)

Reply:
Yes, especially if it's not editable.  If that changed, as this does talk of using this pattern for other things, having a static name and creation managed by the platform will address rbac needs.

Reply:
when we talk of using this pattern for other things, we mean there might be other cases where we add additional key/values to the gate configmap, and require the admin ack them in the ack configmap.

the actual configmap name/namespace would be the same one that's being defined here, not new/additional ones.